flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/ExecutorImpl.java

```diff
@@ -429,4 +453,46 @@ private CompletableFuture<OperationStatusResponseBody> closeOperationAsync(
     private OperationHandle getOperationHandle(Supplier<String> handleSupplier) {
         return new OperationHandle(UUID.fromString(handleSupplier.get()));
     }
+
+    private SqlGatewayRestAPIVersion negotiateVersion() {
+        List<SqlGatewayRestAPIVersion> gatewayVersions =
+                getResponse(
+                                sendRequest(
+                                        GetApiVersionHeaders.getInstance(),
+                                        EmptyMessageParameters.getInstance(),
+                                        EmptyRequestBody.getInstance(),
+                                        // Currently, RestClient always uses the latest REST API
+                                        // version to build the targetUrl. However, it's possible
+                                        // that the client REST API version is higher than the
+                                        // server REST API version. In this case, the gateway will
+                                        // report Not Found Error to notify the client.
+                                        //
+                                        // So, here use the lowest REST API version to get the
+                                        // remote gateway version list and then determine the
+                                        // connection version.
+                                        // TODO: Remove this after the REST Client should allow
+                                        // to build the target URL without API version.
+                                        Collections.min(
+                                                Arrays.stream(SqlGatewayRestAPIVersion.values())
+                                                        .filter(
+                                                                SqlGatewayRestAPIVersion
+                                                                        ::isStableVersion)
+                                                        .collect(Collectors.toList()))))
+                        .getVersions().stream()
+                        .map(SqlGatewayRestAPIVersion::valueOf)
+                        .collect(Collectors.toList());
+        SqlGatewayRestAPIVersion clientVersion = SqlGatewayRestAPIVersion.getDefaultVersion();
+
+        if (gatewayVersions.contains(clientVersion)) {
+            return clientVersion;
+        } else {
+            SqlGatewayRestAPIVersion latestVersion =
+                    RestAPIVersion.getLatestVersion(gatewayVersions);
+            if (latestVersion.equals(SqlGatewayRestAPIVersion.V1)) {
+                throw new SqlExecutionException(
+                        "Currently SQL Client only supports to connect to the REST endpoint whose API version is larger than V1.");
```

Comment:
Nit: suggested by Grammarly
`Currently SQL Client only supports to connect to the REST endpoint whose API version is larger than V1.` 
 => 
`Currently, SQL Client only supports connecting to the REST endpoint with API version larger than V1.`